### ISSUE TITLE
Allow get_dense_feature_matrix() to take a list of species to ensure consistent output size

### ISSUE THREAD

**Issue by Luthaf** [cosmo-epfl/librascal#178]
For ML applications, if the training and test/prediction set do not have the same number of species, `get_dense_feature_matrix` will return different matrix sizes.

This can lead to errors if the test set contains **less** species than the training set, even though this should be a valid case.

And this should be invalid if the test set contains **more** species than the training set.

**Comment by felixmusil** [comment]
This function is not meant for training purposes but more playing around with the coefficients. To handle this use case properly one should provide at the representation level the set of species that need to be considerred and/or as an argument to this function.

**Comment by max-veit** [comment]
Right, in an actual ML application we'd expect users to use `kernels.hh`, which implements the species-aware dot product between properties in C++.

**Comment by felixmusil** [comment]
it makes sense to do that and the simplest way to get this behavior is the expose to python a ´get_dense_feature_matrix ´function that takes a list of atomic numbers as input and checks that it at least contains the species that are present. Note the behavior will need to change depending on the representation (RadialSpectrum, PowerSpectrum, BiSpectrum...).

**Comment by ceriottm** [comment]
Yes. I very much like the idea of being able to access the features in a
structured way from Python.
That way one can generate the feature matrix any way *e likes, and it's
really easy to play around with the coefficients
in a meaningful way.  By exposing this as a function (that has no pretense
of being fast!) we can also deal with e.g.
storing of just half of the coefficients, etc. transparently.

**Comment by max-veit** [comment]
(changed the issue title to reflect that we're now discussing an enhancement, not a bug)

**Comment by Luthaf** [comment]
I would implement this the other way around: instead of having the function take a list of species, it should return a dictionary-like object associating species and the corresponding feature vectors, for all the species it knows about.

**Comment by max-veit** [comment]
But that's what the internal C++ data structure already does, and we agreed that this isn't the most user-friendly way to expose the feature vectors (because they have to write all the logic of species_list_1 ∩ species_list_2, every time they want to do a feature dot product).

This implementation would be okay if we also expose a dot-product function in Python that takes two species-indexed feature matrix dictionaries and builds a kernel matrix out of those (this might already be implemented for C++ features...). That way we parallel the existing C++ implementation in the way it's exposed.

**Comment by Luthaf** [comment]
> But that's what the internal C++ data structure already does, and we agreed that this isn't the most user-friendly way to expose the feature vectors

My main issue with taking a list of species is that the user of the code then has to create this list, when the corresponding information is already known by the c++ side.

> because they have to write all the logic of species_list_1 ∩ species_list_2, every time they want to do a feature dot product

That is exactly what I want to be able to do, in order for me to play with species, and features and representation.

> This implementation would be okay if we also expose a dot-product function in Python that takes two species-indexed feature matrix dictionaries and builds a kernel matrix out of those (this might already be implemented for C++ features...). That way we parallel the existing C++ implementation in the way it's exposed.

This may work, but then we are duplicated c++ functionality in Python.

>  But it does not really give you the best performance gain since you would do several dot products so I feel it defeats a bit your purpose.

I have two purposes here: one is performance related, and can (and should!) be fixed on the C++ side. The other is related to having more freedom in the way I can explore the features, even if the resulting code is slow. My proposal here is mostly related to the second.

> the name of the function should be changed though since it is not dense

yes, of course :smiley:

**Comment by ceriottm** [comment]
> I have two purposes here: one is performance related, and can (and
> should!) be fixed on the C++ side. The other is related to having more
> freedom in the way I can explore the features, even if the resulting code
> is slow. My proposal here is mostly related to the second.
>
I really think this consideration is key: performance should be a C++
concern, here it'd be nice to have the most transparent way to access the
invariant feature entries so that one can play around with them. Best I can
think of is really something that would do (on the python side)

representation.structure[n].center[i].get_features(alpha, alpha', n, n', l)

**Comment by Luthaf** [comment]
> representation.structure[n].center[i].get_features(alpha, alpha', n, n', l)

Getting a single numpy array with all parameters would be nicer I think: 

representation.structure[n].center[i].get_features() => np.array((nalpha, nalpha, n_max, n_max, l_max))

**Comment by max-veit** [comment]
This is part of the "improve Python interface to representations" task (which doesn't seem to have an issue yet, but if you like, we can make this issue into that feature request!)
